Fixed #23 완료 표시하기

paikwiki opened the pull request:
- Task 클래스: 단위테스트 추가
- tasks 테이블: closed_at 칼럼 추가
- Task: $fillable: closed_at 추가
- Task: close(), reopen() 메소드 추가
- TaskController: update()에 완료 여부 처리코드 추가
- ModelFactory: 테스트를 위해 Task 팩토리 추가
- tasks/index.blade.php: 완료된 할일 가운데줄 추가
- tasks/index.blade.php: 열기/닫기 버튼 추가

paikwiki committed:
Fixed #23 완료 표시하기

- Task 클래스: 단위테스트 추가
- tasks 테이블: closed_at 칼럼 추가
- Task: $fillable: closed_at 추가
- Task: close(), reopen() 메소드 추가
- TaskController: update()에 완료 여부 처리코드 추가
- ModelFactory: 테스트를 위해 Task 팩토리 추가
- tasks/index.blade.php: 완료된 할일 가운데줄 추가
- tasks/index.blade.php: 열기/닫기 버튼 추가

paikwiki commented:
PR날리고 나서 한 가지 특이사항을 발견했네요.
edit버튼을 눌러서 수정을 하면 완료했던 일이 미완료로 변경됩니다. 했기 때문에 다시 미완료로 바뀐다고 볼 수 있긴 하지만.. 의도했던 동작이 아니기 때문에 나중에 다시 손을 보겠습니다.

tasks/index.blade.php 부분에 hidden input으로 값을 넘기는 방식도 수정이 필요하고.. Task의 reopen에 불필요한 타임존 설정 코드도 실수로 작성한 걸 PR후에 발견했네요. 여튼 다른 분이 코드 리뷰해주시고 머지 해주시면 제 담당 이슈인 #9 "현황 뷰 추가, 할일 개수 표시(추후 할일/한일 표시할 수 있도록 준비)"도 18일까지 마무리하겠습니다.

smartbos commented on app/Http/Controllers/TaskController.php:
```diff
@@ -58,6 +59,15 @@ public function update(StoreTaskRequest $request, $id)
+        // 완료 여부 표시
+        if( $request->input('closed_at') )
+        {
+            $task->close();
```
$task를 업데이트하려고 프로퍼티에 값을 대입하던 중이니까 

`$task->closed_at = Carbon::now();`

로 하는게 맞을 것 같아요.

현재 코드에서는 71번 라인의 $task->save()가 $task->close()나 $task->reopen()을 덮어쓰는 것 같아요.

paikwiki replied:
다시 한 번 살펴보도록 하겠습니다. 연휴가 너무 빨리 지나갔네요;

smartbos commented on app/Task.php:
```diff
@@ -55,4 +56,17 @@ public function getDdays( $now )
+    /* 완료 표시하기 */
+    public function close()
+    {
+        date_default_timezone_set('Asia/Seoul'); // 한국시간 문제
```
시간 설정은 앱 전체에 동일하게 적용되어야 할테니 아예 `config/app.php` 파일에서 'timezone' 값을 변경하는게 어떨까요?

smartbos commented on database/factories/ModelFactory.php:
```diff
@@ -22,3 +22,12 @@
+$factory->define(App\Task::class, function (Faker\Generator $faker) {
+    return [
+        'user_id' => function() {
+            return App\User::find(1);
```
데이터베이스에 유저가 한 명도 없으면 테스트가 동작하지 않겠네요.
더 안전하게 유저도 한 명 만드는 건 어떨까요?

smartbos replied:
아니면 유저가 존재하는지 여부에 따라 없으면 유저를 만드는 로직을 추가하는 것도 괜찮을 것 같구요.

paikwiki replied:
유저와 관련된 수정도 필요하겠네요. 코멘트 고맙습니다.

paikwiki committed:
Fixed #23 again - 타임존, 완료 표시 기능 수정

- Task.php: close()와 reopen()에서 객체의 프로퍼티 값만 할당
- Task.php: 타임존 설정 코드 삭제(app.php에서 타임존 적용)
- config/app.php: 타임존 설정을 UTC에서 Asia/Seoul로 변경
- ModelFactory.php: Task 생성 팩토리에서 사용자가 없을 경우 사용자를 생성해서 할일을 작성하도록 변경
- TaskTest.php: close()와 reopen() 매소드가 값을 변경하지 않고 값을 할당하기만 하기 때문에 closed_at 값을 직접 할당해서 테스트 실행하도록 변경

paikwiki commented:
수정해서 다시 커밋했는데 이 PR에 커밋이 붙는(?) 거 같네요. 수정내용 확인 부탁드립니다! https://github.com/hongdae-laravel/hong-todo/pull/24/commits/0230ac6089e30bf3825945bcc8fd60b0047ce164
